internal/runner/nomad_manager.go

```diff
@@ -95,69 +134,100 @@ func (m *NomadRunnerManager) SynchronizeRunners(ctx context.Context) error {
 	return err
 }
 
+func (m *NomadRunnerManager) StoreEnvironment(environment ExecutionEnvironment) {
+	m.AbstractManager.StoreEnvironment(environment)
+	m.reloadingEnvironment[environment.ID()] = &sync.Mutex{}
+}
+
+func (m *NomadRunnerManager) DeleteEnvironment(id dto.EnvironmentID) {
+	m.AbstractManager.DeleteEnvironment(id)
+	delete(m.reloadingEnvironment, id)
+}
+
 // Load recovers all runners for all existing environments.
-func (m *NomadRunnerManager) load() error {
+func (m *NomadRunnerManager) load() {
 	newUsedRunners := storage.NewLocalStorage[Runner]()
-	for _, environment := range m.environments.List() {
-		environmentLogger := log.WithField(dto.KeyEnvironmentID, environment.ID().ToString())
-
-		runnerJobs, err := m.apiClient.LoadRunnerJobs(environment.ID())
+	for _, environment := range m.ListEnvironments() {
+		usedRunners, err := m.loadEnvironment(environment)
 		if err != nil {
-			return fmt.Errorf("failed fetching the runner jobs: %w", err)
+			log.WithError(err).WithField(dto.KeyEnvironmentID, environment.ID().ToString()).
+				Warn("Failed loading environment. Skipping ...")
+			continue
 		}
-		for _, job := range runnerJobs {
-			m.loadSingleJob(job, environmentLogger, environment, newUsedRunners)
+		for _, r := range usedRunners.List() {
+			newUsedRunners.Add(r.ID(), r)
 		}
-		err = environment.ApplyPrewarmingPoolSize()
+	}
+
+	m.updateUsedRunners(newUsedRunners, true)
+}
+
+func (m *NomadRunnerManager) loadEnvironment(environment ExecutionEnvironment) (used storage.Storage[Runner], err error) {
```

Comment:
What is the difference between `NomadRunnerManager.loadEnvironment` and `NomadEnvironment.ApplyPrewarmingPoolSize`? Are we creating parallel structures for the same task? Shouldn't it be in the control of the environment to manage its runners?

I did so as the Nomad Environment...
1. has no control over the used runners (which otherwise would not be loaded).
2. cannot create Nomad Runner References (just indirectly by requesting a new runner at Nomad and Nomad notifying the Runner Manager via the Event Stream).
3. the load/recover functionality was historically always directly in the runner manager.

Even though it's not optimal, I would go with it. What do you think?

Reply:
Actually, I don't see where we are duplicating code or behaviour in these two methods. Sure, `NomadRunnerManager.loadEnvironment` is calling `NomadEnvironment.ApplyPrewarmingPoolSize`, but that feels right. Otherwise, and especially because `NomadEnvironment.ApplyPrewarmingPoolSize` is so small, I don't see an issue and would proceed as you suggested.